File: cpp/aln/parser.cpp
#include <stdio.h>
#include "parser.h"

Parser::Parser(void (*handler)(Packet*)) {
    onPacket = handler;
	reset();
}

void Parser::ingestFrameBytes(uint8* in, int sz) {
    for (int i = 0; i < sz; i++) {
		if (in[i] == FRAME_END) {
			acceptPacket();
		} else if (state == STATE_ESCAPED) {
			switch (in[i]) {
			case FRAME_END_T:
				packetBuffer[pBuffIdx++] = FRAME_END;
			case FRAME_ESC_T:
				packetBuffer[pBuffIdx++] = FRAME_ESC;
			}
			state = STATE_BUFFERING;
		} else if (in[i] == FRAME_ESC) {
			state = STATE_ESCAPED;
		} else {
			packetBuffer[pBuffIdx++] = in[i];
		}
	}
}

void Parser::acceptPacket() {
	parsePacket();
	onPacket(&packet);
	reset();
}


// ParsePacket makes a structured packet from an unframed buffer
void Parser::parsePacket() {
	packet.cf = cfHamDecode(readUint16(packetBuffer));
	uint8 offset = CF_FIELD_SIZE; // length of controlFlags
	if (packet.cf&CF_NETSTATE) {
		printf("net\n");
		packet.net = packetBuffer[offset++];
	}
	if (packet.cf&CF_SERVICE) {
		printf("srv\n");
		packet.srvSz = packetBuffer[offset++];
		packet.srv = packetBuffer+offset;
		offset += packet.srvSz;
	}
	if (packet.cf&CF_SRCADDR) {
		printf("src\n");
		uint8 addrSize = packetBuffer[offset++];
		packet.src = packetBuffer+offset;
		offset += addrSize;
	}
	if (packet.cf&CF_DESTADDR) {
		printf("dst, offset:%d\n", offset);
		packet.dstSz = packetBuffer[offset++];
		printf("dst, addrSize:%d\n", packet.dstSz);
		packet.dst = packetBuffer+offset;
		printf("dst:%.*s\n", packet.dstSz, packet.dst);
		offset += packet.dstSz;
		printf("offset:%d\n", offset);
	}
	if (packet.cf&CF_NEXTADDR) {
		printf("nxt\n");
		packet.nxtSz = packetBuffer[offset++];
		packet.nxt = packetBuffer+offset;
		offset += packet.nxtSz;
	}
	if (packet.cf&CF_SEQNUM) {
		printf("seq\n");
		packet.seq = readUint16(packetBuffer+offset);
		offset += SEQNUM_FIELD_SIZE;
	}
	if (packet.cf&CF_ACKBLOCK) {
		printf("ack\n");
		packet.ack = readUint32(packetBuffer+offset);
		offset += ACKBLOCK_FIELD_SIZE;
	}
	if (packet.cf&CF_CONTEXTID) {
		printf("ctx\n");
		packet.ctx = readUint16(packetBuffer+offset);
		offset += CONTEXTID_FIELD_SIZE;
	}
	if (packet.cf&CF_DATATYPE) {
		printf("typ\n");
		packet.typ = packetBuffer[offset++];
	}
	if (packet.cf&CF_DATA) {
		printf("data\n");
		packet.dataSz = readUint16(packetBuffer+offset);
		offset += DATALENGTH_FIELD_SIZE;
		packet.data = packetBuffer+offset;
	}
}

void Parser::reset() {
	packet.clear();
	pBuffIdx = 0;
	state = STATE_BUFFERING;
}

File: nodejs/lib/aln/parser.js
const ByteBuffer = require('bytebuffer')
const { Packet } = require('./packet')

const StateBuffering = 0
const StateEscaped = 1
const End = 0xC0
const Esc = 0xDB
const EndT = 0xDC
const EscT = 0xDD

function toAx25Frame (content) {
  const frame = new ByteBuffer(4096)
  for (let i = 0; i < content.length; i++) {
    const b = content.charCodeAt(i)
    if (b === End) {
      frame.append([Esc])
      frame.append([EndT])
    } else if (b === Esc) {
      frame.append([Esc])
      frame.append([EscT])
    } else {
      frame.append([b])
    }
  }
  frame.append([End])
  const offset = frame.offset
  return frame.reset().toBinary(0, offset)
}

class Parser {
  constructor (onPacket) {
    this.state = StateBuffering
    this.buffer = new ByteBuffer(4096)
    this.onPacket = onPacket
  }

  reset () {
    this.state = StateBuffering
    this.buffer.reset()
  }

  acceptPacket () {
    const offset = this.buffer.offset
    const buf = this.buffer.reset().toBinary(0, offset)
    const p = new Packet(buf)
    this.onPacket(p)
    this.reset()
  }

  readAx25FrameBytes (data, length) {
    for (let i = 0; i < length; i++) {
      const b = data.charCodeAt(i)
      if (b === End) {
        this.acceptPacket()
      } else if (this.state === StateEscaped) {
        if (b === EndT) {
          this.buffer.append([End])
        } else if (b === EscT) {
          this.buffer.append([Esc])
        }
        this.state = StateBuffering
      } else if (b === Esc) {
        this.state = StateEscaped
      } else {
        this.buffer.append([b])
      }
    }
  }
}

module.exports.Parser = Parser
module.exports.toAx25Frame = toAx25Frame
